src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java

```diff
@@ -325,11 +370,50 @@ private static TableReference getTableReference(Configuration conf) throws IOExc
       return Optional.empty();
     }
 
+    private void operationAction(TableReference tableRef) throws InterruptedException {
+      String query = generateQuery(tableRef);
+
+      BigQuery bigquery = BigQueryOptions.getDefaultInstance().getService();
+      QueryJobConfiguration queryConfig = QueryJobConfiguration.newBuilder(query).setUseLegacySql(false).build();
+
+      JobId jobId = JobId.of(UUID.randomUUID().toString());
+      com.google.cloud.bigquery.Job queryJob = bigquery.create(JobInfo.newBuilder(queryConfig).setJobId(jobId).build());
+
+      // Wait for the query to complete.
+      queryJob.waitFor();
+    }
+
+    private String generateQuery(TableReference tableRef) {
+      String criteriaTemplate = "T.%s = S.%s";
+      String sourceTable = temporaryTableReference.getDatasetId() + "." + temporaryTableReference.getTableId();
+      String destinationTable = tableRef.getDatasetId() + "." + tableRef.getTableId();
+      String criteria = tableKeyList.stream().map(s -> String.format(criteriaTemplate, s, s))
+        .collect(Collectors.joining(" AND "));
+      String fieldsForUpdate = tableFieldsList.stream().filter(s -> !tableKeyList.contains(s))
+        .map(s -> String.format(criteriaTemplate, s, s)).collect(Collectors.joining(", "));
+      switch (operation) {
+        case UPDATE:
+          return String.format(UPDATE_QUERY, destinationTable, fieldsForUpdate, sourceTable, criteria);
+        case UPSERT:
+          String insertFields = String.join(", ", tableFieldsList);
+          return String.format(UPSERT_QUERY, destinationTable, sourceTable, criteria, fieldsForUpdate,
+                                insertFields, insertFields);
+        default:
+          return "";
+      }
+    }
+
     private static TableSchema createTableSchemaFromFields(String fieldsJson) throws IOException {
       List<TableFieldSchema> fields = new ArrayList<>();
       JsonParser parser = JacksonFactory.getDefaultInstance().createJsonParser(fieldsJson);
       parser.parseArrayAndClose(fields, TableFieldSchema.class);
       return new TableSchema().setFields(fields);
     }
+
+    private boolean isTableEmpty(TableReference tableRef) throws IOException {
+      Table table = bigQueryHelper.getTable(tableRef);
+      return table.getNumRows().compareTo(BigInteger.ZERO) <= 0 && table.getNumBytes() == 0
```

Comment:
does getNumRows() result in a BQ query or is it just metadata? If it has to run through all the data to do a count then it could be more expensive than just running the update query.

Reply:
Also there is cost implication. Scans cost on BigQuery.

Reply:
getNumRows() is it just metadata.